pedersen/src/product_protocol.rs

```diff
@@ -0,0 +1,557 @@
+
+        cy.serialize_compressed(&mut compressed_bytes).unwrap();
+        transcript.append_point(b"CY", &compressed_bytes[..]);
+
+        cxy.serialize_compressed(&mut compressed_bytes).unwrap();
+        transcript.append_point(b"CXY", &compressed_bytes[..]);
+
+        alpha.serialize_compressed(&mut compressed_bytes).unwrap();
+        transcript.append_point(b"alpha", &compressed_bytes[..]);
+
+        beta.serialize_compressed(&mut compressed_bytes).unwrap();
+        transcript.append_point(b"beta", &compressed_bytes[..]);
+
+        delta.serialize_compressed(&mut compressed_bytes).unwrap();
+        transcript.append_point(b"delta", &compressed_bytes[..]);
+    }
+
+    /// create. This function returns a new product proof for `x`, `y` against 
+    /// `cx`, `cy` and `cxy`.
+    /// # Arguments
+    /// * `transcript` - the transcript object that is modified.
+    /// * `rng` - the RNG that is used to produce the random values. Must be 
+    /// cryptographically secure.
+    /// * `x` - the value that is used to show an opening of `cx`.
+    /// * `y` - the value that is used to show an opening of `cy`.
+    /// * `cx` - the commitment to x that is opened.
+    /// * `cy` - the commitment to y that is opened.
+    /// * `cxy` - the commitment to xy that is opened.
+    pub fn create<T: RngCore + CryptoRng>(
+        transcript: &mut Transcript,
+        rng: &mut T,
+        x: &<P as CurveConfig>::ScalarField,
+        y: &<P as CurveConfig>::ScalarField,
+        cx: &PedersenComm<P>,
+        cy: &PedersenComm<P>,
+        cxy: &PedersenComm<P>,
+    ) -> Self {
+        // This function just creates the intermediary objects and makes the 
+        // proof from those.
+        let inter = Self::create_intermediates(
+            transcript,
+            rng,
+            cx,
+            cy,
+            cxy,
+        );
+
+        // Now call the routine that returns the "challenged" version.
+        // N.B For the sake of compatibility, here we just pass the buffer 
+        // itself.
+        let chal_buf = transcript.challenge_scalar(b"c");
+        Self::create_proof(x, y, &inter, cx, cy, cxy, &chal_buf)
+    }
+
+    /// create_intermediaries. This function returns a new set of intermediaries
+    /// for a new product proof for `x`, `y` against `cx`, `cy` and `cxy`.
+    /// # Arguments
+    /// * `transcript` - the transcript object that is modified.
+    /// * `rng` - the RNG that is used to produce the random values. Must be 
+    /// cryptographically secure.
+    /// * `cx` - the commitment to x that is opened.
+    /// * `cy` - the commitment to y that is opened.
+    /// * `cxy` - the commitment to xy that is opened.
+    pub fn create_intermediates<T: RngCore + CryptoRng>(
+        transcript: &mut Transcript,
+        rng: &mut T,
+        cx: &PedersenComm<P>,
+        cy: &PedersenComm<P>,
+        cxy: &PedersenComm<P>,
+    ) -> ProductProofIntermediate<P> {
```

Comment:
Would it make more sense for this to be a method on the `ProductProofIntermediate` type instead?